src/db.rs

```diff
@@ -260,27 +262,60 @@ impl DbInner {
         }
     }
 
-    async fn replay_wal(&self) -> Result<(), SlateDBError> {
+    async fn replay_wal(self: &Arc<Self>) -> Result<(), SlateDBError> {
         let wal_id_last_compacted = self.state.read().state().core.last_compacted_wal_sst_id;
         let wal_sst_list = self
             .table_store
             .get_wal_sst_list(wal_id_last_compacted)
             .await?;
         let mut last_sst_id = wal_id_last_compacted;
+        let mut wal_tasks = Vec::new();
+        // we use a BTreeMap here to preserve order of sst_id
+        let mut wal_replay_buffers = BTreeMap::new();
+        let mut wal_replay_errors = Vec::new();
+        let semaphore = Arc::new(tokio::sync::Semaphore::new(
+            self.options.maximum_parallel_tasks_limit,
+        ));
+
+        // constructs a list of tokio tasks to asynchronously fetch wal sst
         for sst_id in wal_sst_list {
-            last_sst_id = sst_id;
-            let sst = self.table_store.open_sst(&SsTableId::Wal(sst_id)).await?;
-            let sst_id = match &sst.id {
-                SsTableId::Wal(id) => *id,
-                SsTableId::Compacted(_) => return Err(SlateDBError::InvalidDBState),
-            };
-            let mut iter = SstIterator::new(&sst, self.table_store.clone(), 1, 1, true).await?;
```

Comment:
I was taking a look at existing code and it looks like the SstIterator already has built-in block-level parallelism, why not leverage that instead of fetching the WALs themselves in parallel? That seems like it would be more efficient from a memory perspective and still achieve identical results (cf my other comment)

Reply:
SstIterator only parallelizes reads to a single SST. The active part of the WAL can have 100s of small SSTs if the `flush_interval` is small and the `l0_sst_size_bytes` is large(r). The parallel reads in SstIterator won't help us in this situation since it doesn't span multiple SSTs.

I took a look at `merge_iterator.rs` and it also doesn't seem to have an iterator that can prefetch SSTs when we're reading a bunch of them consecutively. Instead, it's built to handle range queries where we read the SSTs in parallel and merge their items based on whichever key is smaller.

That said, I think it might be better to add a SstPrefetchIterator to `merge_iterator.rs` that receives a list of SST IDs and reads up to N of them in parallel.

Reply:
SstIterator only parallelizes reads to a single SST. The active part of the WAL can have 100s of small SSTs if the `flush_interval` is small and the `l0_sst_size_bytes` is large(r). The parallel reads in SstIterator won't help us in this situation since it doesn't span multiple SSTs.

I took a look at `merge_iterator.rs` and it also doesn't seem to have an iterator that can prefetch SSTs when we're reading a bunch of them consecutively. Instead, it's built to handle range queries where we read the SSTs in parallel and merge their items based on whichever key is smaller.

That said, I think it might be better to add a SstPrefetchIterator to `merge_iterator.rs` that receives a list of SST IDs and reads up to N of them in parallel.

Reply:
Perhaps the easiest way to do this is simply to create N `SstIterator`s with `new_spawn`. This will cause each `SstIterator` to pre-fetch its first `max_fetch_tasks` blocks immediately, and in parallel. When the first `SstIterator` is exhausted, we push a new SstIterator (with `new_spawn`) to the back of the list if iterators, and pop the next off the front to iterate.

I'm trying to figure out the simplest thing we can do that gets us some measure of parallelism. I think this approach will work well. For large SSTs, the `SstIterator` will only prefetch a subset of blocks, keeping memory pressure down. For small SSTs, it will result in pre-fetching the first N SSTs in parallel (assuming they're each only a few blocks in size). This should reduce startup latency.

Reply:
thanks for pointing this out @agavra and starting this discussion, I hadn't looked into `SstIterator`'s code. I will wait for Rohan to chime in and then start making changes.

Reply:
> SstIterator only parallelizes reads to a single SST. The active part of the WAL can have 100s of small SSTs if the flush_interval is small and the l0_sst_size_bytes is large(r). The parallel reads in SstIterator won't help us in this situation since it doesn't span multiple SSTs.

Yeah, it probably won't help. I wrote that with compaction in mind, where you're reading whole SSTs that are quite large (64MB at least, and typically 100s of MBs - 1GB). The common case here is that there are lots of small SSTs to replay.

The N SSTIterators approach @criccomini outlined makes sense to me. We can create each iterator with 1 fetch task that fetches up to 256 blocks (1MB). I think for most cases that'll cause the whole SST to be read. I think we can just do that inline in the replay code rather than defining a SequentialIterator class. I don't think there are other use cases, and everywhere else in the code base we want the iterators to return keys in sorted order.

Reply:
SGTM!